table.go

```diff
@@ -72,13 +82,67 @@ func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerst
 		PeerAdded:   func(peer.ID) {},
 
 		maxLastSuccessfulOutboundThreshold: maxLastSuccessfulOutboundThreshold,
+		rtRefreshInterval:                  rtRefreshInterval,
+	}
+
+	if peerPingFnc == nil {
+		return nil, errors.New("peerPing func is nil")
 	}
+	rt.peerPingFnc = peerPingFnc
+
+	if peerConnectednessFnc == nil {
+		return nil, errors.New("peerConnectednesss func is nil")
+	}
+	rt.peerConnectednessFnc = peerConnectednessFnc
 
 	rt.ctx, rt.ctxCancel = context.WithCancel(context.Background())
+	go rt.background()
 
 	return rt, nil
 }
 
+func (rt *RoutingTable) background() {
+	tickr := time.NewTicker(rt.rtRefreshInterval / 3)
+	defer tickr.Stop()
+
+	for {
+		select {
+		case <-tickr.C:
+			// get all peers in the routing table
+			rt.tabLock.RLock()
+			var peers []PeerInfo
+			for _, b := range rt.buckets {
+				peers = append(peers, b.peers()...)
+			}
+			rt.tabLock.RUnlock()
+
+			// start going through them
+			for _, ps := range peers {
+				// ping the peer if it's due for a ping and evict it if the ping fails
+				if time.Since(ps.lastSuccessfulOutboundQuery) > (rt.rtRefreshInterval / 3) {
+					livelinessCtx, cancel := context.WithTimeout(rt.ctx, 10*time.Second)
+					if err := rt.peerPingFnc(livelinessCtx, ps.Id); err != nil {
+						log.Debugf("failed to ping peer=%s, got error=%s, evicting it from the RT", ps.Id, err)
+						// below lock and connectedness check helps prevents the following race:
+						// we ping a peer and realize it's disconnected -> we somehow get a connection to the peer but after
+						// our ping and before we remove it from the RT -> which means rt.TryAdd() will not do anything ->
+						// we remove it from the RT thus losing a connected peer.
+						rt.tabLock.Lock()
+						if !rt.peerConnectednessFnc(ps.Id) {
+							rt.removePeer(ps.Id)
+						}
+						rt.tabLock.Unlock()
```

Comment:
This is racy as well. We could have sent them a query in the meanwhile which would lead to us removing them even though they're useful (just overloaded).

Perhaps we can check if the peers `ps.lastSuccessfulOutboundQuery` > time before we started the ping?